lib/netplay/port_mapping_manager.h

```diff
@@ -195,15 +221,36 @@ class PortMappingManager
 	static void scheduleCallbacksOnMainThread(CallbacksPerRequest cbReqMap);
 
 	// Must only be called internally, and if mtx_ lock is held!
-	PortMappingAsyncRequestPtr active_request_for_id_internal(int id) const;
-	static void resolve_success_internal(PortMappingAsyncRequestPtr req, std::string externalIp, uint16_t externalPort);
-	static void resolve_failure_internal(PortMappingAsyncRequestPtr req, PortMappingDiscoveryStatus failureStatus);
+	PortMappingAsyncRequestPtr active_request_for_id_internal(PortMappingImpl::Type type, int id) const;
+	static bool resolve_success_internal(PortMappingAsyncRequestPtr req, std::string externalIp, uint16_t externalPort);
+	bool resolve_failure_internal(PortMappingAsyncRequestPtr req, PortMappingDiscoveryStatus failureStatus);
 
+	struct ImplMappingId
+	{
+		PortMappingImpl::Type implType;
+		PortMappingAsyncRequest::MappingId mappingId;
+
+		bool operator==(const ImplMappingId &other) const
+		{
+			return (implType == other.implType
+				  && mappingId == other.mappingId);
+		}
+	};
+	struct ImplMappingIdHasher
+	{
+		std::size_t operator()(const ImplMappingId& k) const
+		{
+			return std::hash<int>()(static_cast<int>(k.implType)) ^ std::hash<int>()(k.mappingId);
+		}
+	};
 
-	std::unordered_map<PortMappingAsyncRequest::MappingId, PortMappingAsyncRequestPtr> activeDiscoveries_;
+	std::unordered_map<ImplMappingId, PortMappingAsyncRequestPtr, ImplMappingIdHasher> activeDiscoveries_;
 	std::unordered_map<PortWithProtocol, PortMappingAsyncRequestPtr> successfulRequests_;
 
 	bool isInit_ = false;
+	std::vector<std::shared_ptr<PortMappingImpl>> loadedImpls;
+	std::shared_ptr<PortMappingImpl> currentImpl_ = nullptr;
```

Comment:
```suggestion
	PortMappingImpl* currentImpl_ = nullptr;
```
This should point into `loadedImpls_`.